tiledb/sm/stats/global_stats.cc

```diff
@@ -89,8 +89,14 @@ void GlobalStats::set_enabled(bool enabled) {
 
 void GlobalStats::reset() {
   std::unique_lock<std::mutex> ul(mtx_);
-  for (auto& register_stat : registered_stats_) {
-    register_stat->reset();
+  for (auto register_stat = registered_stats_.begin(); register_stat != registered_stats_.end();) {
+    // Reset the stats, -not- the shared_ptr!
+    (*register_stat)->reset();
+    if (register_stat->use_count() <= 1) {
```

Comment:
This is replicating the feature of `weak_ptr` that self-destroys objects under the same conditions. Changing the declaration of `registered_stats_` would yield better memory performance by allowing its contents to destruct promptly, rather than waiting for this function `reset()` to run. This clean-up pass would still be necessary to destroy the `weak_ptr`, but there's much less memory consumption there than with the `Stats` object itself.
```
  /** The aggregated stats. */
  std::list<weak_ptr<stats::Stats>> registered_stats_;
```

Reply:
> Note this is a potentially breaking change in that stats associated with the registered object will now have to be obtained ('dump()'ed) before that object is destroyed, whereas before it would still be available beyond the life time of the registrant.

Specifically, IIUC this means that the stats associated with a context will now be removed (effectively zero'd) when the context is deallocated. Whereas currently, that happens only when `stats_reset` is called -- via zero'ing of the counts in each `Stats` object (but without actually destroying the `Stats` object -- the crux of the problem this PR addresses).

That's a fairly large change to the established/user-understood usage pattern of the current stats API, which is that all stats are aggregated from `tiledb_stats_enable` to `tiledb_stats_reset` (or disable). eg in python we have:
```
tiledb.stats_enable()
# do stuff, which may include creating and destroying contexts
tiledb.stats_dump() # output includes recorded stats from all contexts, dead or alive
```
This is flawed in various ways, but aggregation is IMO the least surprising behavior given the current global-only start/stop/reset access to stats. I think we should accept the overhead of the shared_ptr in order to keep the current behavior, but fix the leak by removing upon `stats_reset` as was originally done.

(an alternative approach to maintaining the current behavior might be to manually aggregate/sum the key:count pairs into a single long-lived object when a context is destroyed. doing that kind of extracurricular work during object destruction feels sketchy, and moreso if it has to take a mutex to avoid racing)